bot/exts/utils/snekbox/_cog.py

```diff
@@ -289,6 +290,69 @@ async def format_output(
 
         return output, paste_link
 
+    async def format_file_text(self, text_files: list[FileAttachment], output: str) -> str:
+        # Inline until budget, then upload to paste service
+        # Budget is shared with stdout, so subtract what we've already used
+        budget_lines = MAX_OUTPUT_BLOCK_LINES - (output.count("\n") + 1)
+        budget_chars = MAX_OUTPUT_BLOCK_CHARS - len(output)
+        msg = ""
+
+        for file in text_files:
+            file_text = file.content.decode("utf-8", errors="replace") or "[Empty]"
+            # Override to always allow 1 line and <= 50 chars, since this is less than a link
+            if len(file_text) <= 50 and not file_text.count("\n"):
+                msg += f"\n`{file.name}`\n```\n{file_text}\n```"
+            # otherwise, use budget
+            else:
+                format_text, link_text = await self.format_output(
+                    file_text,
+                    budget_lines,
+                    budget_chars,
+                    line_nums=False,
+                    output_default="[Empty]"
+                )
+                # With any link, use it (don't use budget)
+                if link_text:
+                    msg += f"\n`{file.name}`\n{link_text}"
+                else:
+                    msg += f"\n`{file.name}`\n```\n{format_text}\n```"
+                    budget_lines -= format_text.count("\n") + 1
+                    budget_chars -= len(file_text)
+
+        return msg
+
+    def format_blocked_extensions(self, blocked: list[FileAttachment]) -> str:
+        # Sort by length and then lexicographically to fit as many as possible before truncating.
+        blocked_sorted = sorted(set(f.suffix for f in blocked),  key=lambda e: (len(e), e))
+
+        # Only no extension
+        if len(blocked_sorted) == 1 and blocked_sorted[0] == "":
+            blocked_msg = "Files with no extension can't be uploaded."
+        # Both
+        elif "" in blocked_sorted:
+            blocked_str = self.join_blocked_extensions(ext for ext in blocked_sorted if ext)
+            blocked_msg = (
+                f"Files with no extension or disallowed extensions can't be uploaded: **{blocked_str}**"
+            )
+        else:
+            blocked_str = self.join_blocked_extensions(blocked_sorted)
+            blocked_msg = f"Files with disallowed extensions can't be uploaded: **{blocked_str}**"
```

Comment:
If all extensions are too long then it will just show `...`. I think this is fine but it's also possible to completely omit `: **{blocked_str}**"` in such case.